src/main/java/org/folio/service/inventory/InventoryBindingManager.java

```diff
@@ -0,0 +1,90 @@
+package org.folio.service.inventory;
+
+import static org.folio.service.inventory.InventoryUtils.INVENTORY_LOOKUP_ENDPOINTS;
+import static org.folio.service.inventory.InventoryUtils.REQUESTS;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.okapi.common.GenericCompositeFuture;
+import org.folio.rest.core.RestClient;
+import org.folio.rest.core.models.RequestContext;
+import org.folio.rest.core.models.RequestEntry;
+import org.folio.rest.jaxrs.model.Piece;
+import org.folio.service.CirculationRequestsRetriever;
+
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonObject;
+
+public class InventoryBindingManager {
+
+  private static final Logger logger = LogManager.getLogger(InventoryBindingManager.class);
+
+  private static final String REQUEST_CANCEL_MESSAGE = "Cancelling Request with id='{}'";
+  private static final String REQUEST_MOVE_MESSAGE = "Moving Request with id='{}' to item with id='{}'";
+  private static final String REQUEST_ENDPOINT = INVENTORY_LOOKUP_ENDPOINTS.get(REQUESTS) + "/%s";
+  private static final String REQUEST_MOVE_ENDPOINT = REQUEST_ENDPOINT + "/move";
+  private static final String REQUEST_JSON_KEY = "destinationItemId";
+
+  private final CirculationRequestsRetriever circulationRequestsRetriever;
+  private final RestClient restClient;
+
+  public InventoryBindingManager(RestClient restClient, CirculationRequestsRetriever circulationRequestsRetriever) {
+    this.restClient = restClient;
+    this.circulationRequestsRetriever = circulationRequestsRetriever;
+  }
+
+  public Future<List<Piece>> getPiecesWithActiveRequests(List<Piece> pieces, RequestContext requestContext) {
+    if (logger.isDebugEnabled()) {
+      logger.debug("Filtering pieces with active requests: {}", pieces.stream().map(Piece::getId).toList());
+    }
+    var futures = pieces.stream()
+      .map(piece -> getPieceWithRequests(piece, requestContext))
+      .collect(Collectors.toList());
+    return GenericCompositeFuture.all(futures)
```

Comment:
here we can try optimize, because now if we have 10 pieces, we will do 10 calls.
We can introduce new method in CirculationRequestsRetriever to fetch requests by multiple item ids. Interesting do we have in response itemId to map number of requests to itemId.
To avoid very long GET queries we use pattern to split by 15 elements, here is releated constant - MAX_IDS_FOR_GET_RQ_15

Reply:
I believe now it should work correctly the way you explained